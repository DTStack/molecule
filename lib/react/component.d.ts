import { GlobalEvent } from 'mo/common/event';
export interface IComponent<S = any> {
    setState(values: S, callback?: (prevState: S, nextState: S) => void): void;
    render(nextState?: S): void;
    onUpdateState(callback: (prevState: S, nextState: S) => void): void;
    getState(): S;
    /**
     * Subscribe the component event
     * @param name
     * @param callback
     */
    onEvent(name: any, callback: any): void;
}
export declare abstract class Component<S = any> extends GlobalEvent implements IComponent<S> {
    protected abstract state: S;
    private _event;
    constructor();
    /**
     * Set the state values, and notify the view component to re render
     * @param values update target state values
     */
    setState(values: S, callback?: (prevState: S, nextState: S) => void): void;
    /**
     * Initiative notify the component to render the view by the state
     * @param nextState
     */
    render(nextState?: S): void;
    onUpdateState(callback: (prevState: S, nextState: S) => void): void;
    getState(): S;
    onEvent(name: any, callback: any): void;
}
