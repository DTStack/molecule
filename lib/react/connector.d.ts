import * as React from 'react';
import { IComponent } from './component';
import { Controller } from './controller';
export declare type ServiceObject = {
    [index: string]: IComponent;
};
export declare type ControllerObject = {
    [index: string]: Controller;
};
export declare function connect(Service: IComponent | ServiceObject, View: React.ComponentType<any>, Controller?: Controller | ControllerObject, watchFiled?: object): {
    new (props: any): {
        state: {
            lastUpdated: number;
        };
        componentDidMount(): void;
        onChange(prevState: any, nextState: any): void;
        update: () => void;
        getServiceState(): {};
        render(): JSX.Element;
        context: any;
        setState<K extends never>(state: {} | ((prevState: Readonly<{}>, props: Readonly<{}>) => {} | Pick<{}, K> | null) | Pick<{}, K> | null, callback?: (() => void) | undefined): void;
        forceUpdate(callback?: (() => void) | undefined): void;
        readonly props: Readonly<{}> & Readonly<{
            children?: React.ReactNode;
        }>;
        refs: {
            [key: string]: React.ReactInstance;
        };
        shouldComponentUpdate?(nextProps: Readonly<{}>, nextState: Readonly<{}>, nextContext: any): boolean;
        componentWillUnmount?(): void;
        componentDidCatch?(error: Error, errorInfo: React.ErrorInfo): void;
        getSnapshotBeforeUpdate?(prevProps: Readonly<{}>, prevState: Readonly<{}>): any;
        componentDidUpdate?(prevProps: Readonly<{}>, prevState: Readonly<{}>, snapshot?: any): void;
        componentWillMount?(): void;
        UNSAFE_componentWillMount?(): void;
        componentWillReceiveProps?(nextProps: Readonly<{}>, nextContext: any): void;
        UNSAFE_componentWillReceiveProps?(nextProps: Readonly<{}>, nextContext: any): void;
        componentWillUpdate?(nextProps: Readonly<{}>, nextState: Readonly<{}>, nextContext: any): void;
        UNSAFE_componentWillUpdate?(nextProps: Readonly<{}>, nextState: Readonly<{}>, nextContext: any): void;
    };
    contextType?: React.Context<any> | undefined;
};
