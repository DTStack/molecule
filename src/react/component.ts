import { cloneDeep } from 'lodash';
import { EventEmitter, GlobalEvent } from 'mo/common/event';

export enum ComponentEvents {
    Update = 'Component.Update',
}

export interface IComponent<S = any> {
    /**
     * Set the Component state
     * @param values The next values of state
     * @param callback calling after setState
     */
    setState(values: S, callback?: (prevState: S, nextState: S) => void): void;
    /**
     * Trigger the Component update event
     * @param nextState
     */
    render(nextState?: S): void;
    /**
     * Listen to the Component state update event
     * @param listener
     */
    onUpdateState(listener: (prevState: S, nextState: S) => void): void;
    /**
     * Remove the Component update event listening, default is remove all,
     * also you can remove one by pass the listener
     */
    removeOnUpdateState(listener?: Function): void;
    /**
     * Force to update the Component
     */
    forceUpdate(): void;
    /**
     * Get the Component state
     */
    getState(): S;
}

export abstract class Component<S = any>
    extends GlobalEvent
    implements IComponent<S>
{
    protected abstract state: S;
    private _event: EventEmitter;

    constructor() {
        super();
        this._event = new EventEmitter();
    }

    /**
     * Set the state values, and notify the view component to re render
     * @param values update target state values
     */
    public setState(
        values: Partial<S>,
        callback?: (prevState: S, nextState: S) => void
    ) {
        const nextState = Object.assign(this.state, values);
        this.render(nextState);
        callback?.(this.state, nextState);
    }

    /**
     * Initiative notify the component to render the view by the state
     * @param nextState
     */
    public render(nextState?: S) {
        this._event.emit(ComponentEvents.Update, this.state, nextState);
    }

    public onUpdateState(listener: (prevState: S, nextState: S) => void) {
        this._event.subscribe(ComponentEvents.Update, listener);
    }

    public removeOnUpdateState(listener?: Function): void {
        this._event.unsubscribe(ComponentEvents.Update, listener);
    }

    public forceUpdate() {
        this.setState(cloneDeep(this.state));
    }

    public getState(): S {
        return this.state;
    }
}
