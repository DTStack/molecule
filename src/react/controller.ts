import { GlobalEvent } from 'mo/common/event';

export abstract class Controller extends GlobalEvent {
    public abstract initView(): void;
}
