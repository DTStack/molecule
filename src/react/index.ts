export * from './component';
export * from './helper';
export * from './connector';
export * from './controller';
