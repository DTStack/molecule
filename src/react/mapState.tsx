import React, { Component } from 'react';
import Logger from 'mo/common/logger';

/**
 * Mapping the state to the component
 * TODO support mapping service method to the component.
 * @param WrappedComponent The component will be wrapped
 * @param state The state you want to injected, notice the state must be an observable object
 * @param subscribes The events of your subscribe, it used to trigger the state re render
 */
export function mapState<S, T>(
    WrappedComponent: React.ComponentType<S>,
    state: S,
    actions?: T
) {
    return class StateProvider extends Component {
        state: { lastUpdated: number };
        constructor(props) {
            super(props);
            this.onChange = this.onChange.bind(this);
            this.state = {
                lastUpdated: Date.now(),
            };
        }

        componentDidMount() {
            // There is no declare state parameter as IObservable type, so must convert to any type.
            if ((state as any).subscribe) {
                (state as any).subscribe(this.onChange);
            } else {
                Logger.error(
                    'The state parameter of mapState must be an observable object.'
                );
            }
        }
        /**
         * TODO: Performance optimize, now whatever any properties changed in target,
         * there always be trigger the onChange event, so need a compare operation.
         * @param nextState changed data
         */
        onChange(nextState: S) {
            Logger.info(nextState, state);
            this.setState({
                lastUpdated: Date.now(),
            });
        }

        render() {
            return (
                <WrappedComponent
                    {...this.state}
                    {...state}
                    {...this.props}
                    {...actions}
                />
            );
        }
    };
}
