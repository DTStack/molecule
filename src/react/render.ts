import * as ReactDOM from 'react-dom';
import { createRoot, Root } from 'react-dom/client';

const reactdom: typeof ReactDOM & {
    createRoot?: typeof createRoot;
    __SECRET_INTERNALS_DO_NOT_USE_OR_YOU_WILL_BE_FIRED?: {
        usingClientEntryPoint?: boolean;
    };
} = {
    ...ReactDOM,
};

let nextReactDOMClient: typeof createRoot | undefined;

export const renderedSign = Symbol('__molecule__root');

if (ReactDOM.version.startsWith('18')) {
    nextReactDOMClient = reactdom.createRoot;
}

function toggleWarning(skip: boolean) {
    const { __SECRET_INTERNALS_DO_NOT_USE_OR_YOU_WILL_BE_FIRED } = reactdom;

    if (
        __SECRET_INTERNALS_DO_NOT_USE_OR_YOU_WILL_BE_FIRED &&
        typeof __SECRET_INTERNALS_DO_NOT_USE_OR_YOU_WILL_BE_FIRED === 'object'
    ) {
        __SECRET_INTERNALS_DO_NOT_USE_OR_YOU_WILL_BE_FIRED.usingClientEntryPoint =
            skip;
    }
}

export const render = (
    node: JSX.Element,
    container: HTMLElement & {
        [renderedSign]?: Root;
    }
) => {
    if (nextReactDOMClient) {
        toggleWarning(true);
        const root = container[renderedSign] || nextReactDOMClient(container);
        toggleWarning(false);
        root.render(node);

        container[renderedSign] = root;
        return;
    }

    reactdom.render(node, container);
};

export const unmout = (
    container: HTMLElement & {
        [renderedSign]?: Root;
    }
) => {
    if (nextReactDOMClient) {
        container[renderedSign]?.unmount();
        Reflect.deleteProperty(container, renderedSign);
        return true;
    }

    return reactdom.unmountComponentAtNode(container);
};
