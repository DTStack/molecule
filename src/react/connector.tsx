import 'reflect-metadata';
import { container } from 'tsyringe';
import React, { Component } from 'react';
import Logger from 'mo/common/logger';
import { IComponent } from './component';
import { Controller } from './controller';

export type ServiceObject = {
    [index: string]: IComponent;
};

export type ControllerObject = {
    [index: string]: Controller;
};

export function connect<T = any>(
    Service: IComponent | ServiceObject,
    View: React.ComponentType<any>,
    Controller?: Controller | ControllerObject
): React.ComponentType<T> {
    return class Connector extends Component<T, any> {
        state: { lastUpdated: number };
        private _isMounted = false;
        constructor(props) {
            super(props);
            this.onChange = this.onChange.bind(this);
            this.state = {
                lastUpdated: Date.now(),
            };
        }

        componentDidMount() {
            this._isMounted = true;
            this.handleService((service) => {
                service.onUpdateState(this.onChange);
            });
        }

        componentWillUnmount() {
            this._isMounted = false;
            this.handleService((service) => {
                service.removeOnUpdateState(this.onChange);
            });
        }

        onChange(prevState, nextState) {
            Logger.info(prevState, nextState, (container as any)._registry);
            this.update();
        }

        update = () => {
            if (this._isMounted) {
                this.setState({
                    lastUpdated: Date.now(),
                });
            }
        };

        getServiceState() {
            const target = {};
            this.handleService((service, prop) => {
                if (prop) {
                    Object.assign(target, {
                        [prop]: { ...service.getState() },
                    });
                } else {
                    Object.assign(target, { ...service.getState() });
                }
            });
            return target;
        }

        handleService(callback: (service: IComponent, prop?: string) => void) {
            if (this.isValidService(Service as IComponent)) {
                callback(Service as IComponent);
            } else {
                for (const name in Service) {
                    if (name) {
                        const service: IComponent = Service[name];
                        if (this.isValidService(service)) {
                            callback(service, name);
                        }
                    }
                }
            }
        }

        getControllerProperties() {
            const target = {};
            if (!Controller) {
                return target;
            }

            this.handleController((controller: Controller, prop) => {
                if (prop) {
                    Object.assign(target, {
                        [prop]: this.getObjectPublicProperties(controller),
                    });
                } else {
                    Object.assign(
                        target,
                        this.getObjectPublicProperties(controller)
                    );
                }
            });
            return target;
        }

        handleController(
            callback: (controller: Controller, prop?: string) => void
        ) {
            if (this.isValidController(Controller as Controller)) {
                callback(Controller as Controller);
            } else {
                for (const name in Controller) {
                    if (name) {
                        const controller: Controller = Controller[name];
                        if (this.isValidController(controller)) {
                            callback(controller, name);
                        }
                    }
                }
            }
        }

        render() {
            return (
                <View
                    {...this.state}
                    {...this.props}
                    {...this.getServiceState()}
                    {...this.getControllerProperties()}
                />
            );
        }

        private isValidService(service: IComponent) {
            return typeof service.onUpdateState === 'function';
        }

        private isValidController(controller: Controller) {
            return typeof controller.initView === 'function';
        }

        private getObjectPublicProperties(obj: Controller) {
            const keys = Object.keys(obj);
            const result = {};
            keys.forEach((key) => {
                // Filter the Service, Controller instances and private properties which start with '_'
                // TODO There need a better way to filter the private properties of Object,
                // maybe we can make use of the # identifier in future.
                if (
                    !key.endsWith('Service') &&
                    !key.endsWith('Controller') &&
                    !key.startsWith('_')
                ) {
                    result[key] = obj[key];
                }
            });
            return result;
        }
    };
}
