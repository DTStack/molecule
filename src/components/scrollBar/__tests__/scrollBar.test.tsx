import React from 'react';
import { cleanup, fireEvent, render, waitFor } from '@testing-library/react';
import { ScrollBar, DirectionKind, IScrollRef } from '../index';
import {
    scrollBarClassName,
    scrollBarThumbClassName,
    scrollBarTrackClassName,
    scrollBarTrackHiddenClassName,
} from '../base';
import '@testing-library/jest-dom';

const length = 200;

// to make sure the ScrollBar component not be mocked by global
jest.mock('mo/components/scrollBar', () => {
    const originalModule = jest.requireActual('mo/components/scrollBar');
    return {
        ...originalModule,
    };
});

describe('The ScrollBar Component', () => {
    beforeEach(() => {
        // to make sure each ScrollBar is supported to scrolled
        Object.defineProperty(HTMLElement.prototype, 'scrollHeight', {
            configurable: true,
            value: 500,
        });
    });

    let original;
    beforeEach(() => {
        original = HTMLElement.prototype.getBoundingClientRect;
        // @ts-ignore
        HTMLElement.prototype.getBoundingClientRect = jest.fn(() => ({
            x: 0,
            y: 0,
            width: 0,
            height: 0,
        }));

        HTMLElement.prototype.scrollTo = jest.fn();
    });

    let originalRFA;
    beforeEach(() => {
        originalRFA = window.requestAnimationFrame;
        window.requestAnimationFrame = jest.fn((cb) => {
            cb(1);
            return 1;
        });
    });

    afterEach(cleanup);
    afterEach(() => {
        HTMLElement.prototype.getBoundingClientRect = original;
        window.requestAnimationFrame = originalRFA;
    });

    test('Match Snapshot', () => {
        const { asFragment } = render(
            <ScrollBar>
                {new Array(length).fill(1).map((_, idx) => (
                    <div key={idx}>{idx}</div>
                ))}
            </ScrollBar>
        );

        expect(asFragment()).toMatchSnapshot();
    });

    test('Match Horizontal Snapshot', () => {
        const { asFragment } = render(
            <ScrollBar direction={DirectionKind.horizontal}>
                {new Array(length).fill(1).map((_, idx) => (
                    <div key={idx}>{idx}</div>
                ))}
            </ScrollBar>
        );

        expect(asFragment()).toMatchSnapshot();
    });

    test('Should support click track', () => {
        const scrollFn = jest.fn();
        const { container } = render(
            <ScrollBar
                style={{ height: 200, width: 100 }}
                inactiveHidden={false}
                onScroll={scrollFn}
                onScrollStart={scrollFn}
                onScrollEnd={scrollFn}
            >
                {new Array(length).fill(1).map((_, idx) => (
                    <div key={idx}>{idx}</div>
                ))}
            </ScrollBar>
        );

        const track = container.querySelector<HTMLDivElement>(
            `.${scrollBarTrackClassName}`
        );

        fireEvent.click(track!, { clientY: 100 });

        expect(scrollFn).toBeCalledTimes(3);
    });

    test('Should support wheel event on wrapper', () => {
        const scrollFn = jest.fn();
        const { container } = render(
            <ScrollBar
                style={{ height: 200, width: 100 }}
                inactiveHidden={false}
                onScroll={scrollFn}
                onScrollStart={scrollFn}
                onScrollEnd={scrollFn}
            >
                {new Array(length).fill(1).map((_, idx) => (
                    <div key={idx}>{idx}</div>
                ))}
            </ScrollBar>
        );

        const wrapper = container.querySelector(`.${scrollBarClassName}`);

        fireEvent.wheel(wrapper!);

        waitFor(() => {
            expect(scrollFn).toBeCalledTimes(3);
        });
    });

    test('Should support mousemove', () => {
        const scrollFn = jest.fn();
        const { container } = render(
            <ScrollBar
                style={{ height: 200, width: 100 }}
                inactiveHidden={false}
                onScroll={scrollFn}
                onScrollStart={scrollFn}
                onScrollEnd={scrollFn}
            >
                {new Array(length).fill(1).map((_, idx) => (
                    <div key={idx}>{idx}</div>
                ))}
            </ScrollBar>
        );

        const thumb = container.querySelector(`.${scrollBarThumbClassName}`);

        fireEvent.mouseDown(thumb!);

        expect(scrollFn).toBeCalledTimes(1);
        // mousemove would disable the use-select
        expect(document.body.style.userSelect).toBe('none');

        fireEvent.mouseMove(window);
        expect(scrollFn).toBeCalledTimes(2);

        fireEvent.mouseUp(window);
        expect(scrollFn).toBeCalledTimes(3);
        // reset the user-select after mouseup
        expect(document.body.style.getPropertyValue('user-select')).toBe('');
    });

    test('Should visible when mouse over', () => {
        const { container } = render(
            <ScrollBar style={{ height: 200, width: 100 }}>
                {new Array(length).fill(1).map((_, idx) => (
                    <div key={idx}>{idx}</div>
                ))}
            </ScrollBar>
        );

        const track = container.querySelector(`.${scrollBarTrackClassName}`);
        expect(track?.className).toContain(scrollBarTrackHiddenClassName);

        const wrapper = container.querySelector(`.${scrollBarClassName}`)!;

        fireEvent.mouseEnter(wrapper);

        waitFor(() => {
            expect(track?.className).not.toContain(
                scrollBarTrackHiddenClassName
            );
        });

        fireEvent.mouseLeave(wrapper);

        waitFor(() => {
            expect(track?.className).toContain(scrollBarTrackHiddenClassName);
        });
    });

    test('Should support to get instance function from ref', () => {
        const ref = React.createRef<IScrollRef>();

        (HTMLElement.prototype.getBoundingClientRect as jest.Mock)
            .mockClear()
            .mockImplementationOnce(() => ({
                height: 100,
            }))
            .mockImplementationOnce(() => ({
                height: 200,
            }))
            .mockImplementationOnce(() => ({
                height: 100,
            }));

        const { container } = render(
            <ScrollBar
                ref={ref}
                style={{ height: 200, width: 100 }}
                inactiveHidden={false}
            >
                {new Array(length).fill(1).map((_, idx) => (
                    <div key={idx}>{idx}</div>
                ))}
            </ScrollBar>
        );

        waitFor(() => {
            expect(ref.current?.scrollTo).toBeInstanceOf(Function);
        });

        ref.current?.scrollTo(100);

        expect(
            container.querySelector<HTMLDivElement>(
                `.${scrollBarThumbClassName}`
            )?.style.top
        ).toBe('20px');
    });
});
