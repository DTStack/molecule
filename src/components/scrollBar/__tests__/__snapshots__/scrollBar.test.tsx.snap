// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The ScrollBar Component Match Horizontal Snapshot 1`] = `
<DocumentFragment>
  <div
    class="mo-scrollBar__wrapper"
  >
    <div
      class="mo-scrollBar__container mo-scrollBar__container--horizontal"
    >
      <div>
        0
      </div>
      <div>
        1
      </div>
      <div>
        2
      </div>
      <div>
        3
      </div>
      <div>
        4
      </div>
      <div>
        5
      </div>
      <div>
        6
      </div>
      <div>
        7
      </div>
      <div>
        8
      </div>
      <div>
        9
      </div>
      <div>
        10
      </div>
      <div>
        11
      </div>
      <div>
        12
      </div>
      <div>
        13
      </div>
      <div>
        14
      </div>
      <div>
        15
      </div>
      <div>
        16
      </div>
      <div>
        17
      </div>
      <div>
        18
      </div>
      <div>
        19
      </div>
      <div>
        20
      </div>
      <div>
        21
      </div>
      <div>
        22
      </div>
      <div>
        23
      </div>
      <div>
        24
      </div>
      <div>
        25
      </div>
      <div>
        26
      </div>
      <div>
        27
      </div>
      <div>
        28
      </div>
      <div>
        29
      </div>
      <div>
        30
      </div>
      <div>
        31
      </div>
      <div>
        32
      </div>
      <div>
        33
      </div>
      <div>
        34
      </div>
      <div>
        35
      </div>
      <div>
        36
      </div>
      <div>
        37
      </div>
      <div>
        38
      </div>
      <div>
        39
      </div>
      <div>
        40
      </div>
      <div>
        41
      </div>
      <div>
        42
      </div>
      <div>
        43
      </div>
      <div>
        44
      </div>
      <div>
        45
      </div>
      <div>
        46
      </div>
      <div>
        47
      </div>
      <div>
        48
      </div>
      <div>
        49
      </div>
      <div>
        50
      </div>
      <div>
        51
      </div>
      <div>
        52
      </div>
      <div>
        53
      </div>
      <div>
        54
      </div>
      <div>
        55
      </div>
      <div>
        56
      </div>
      <div>
        57
      </div>
      <div>
        58
      </div>
      <div>
        59
      </div>
      <div>
        60
      </div>
      <div>
        61
      </div>
      <div>
        62
      </div>
      <div>
        63
      </div>
      <div>
        64
      </div>
      <div>
        65
      </div>
      <div>
        66
      </div>
      <div>
        67
      </div>
      <div>
        68
      </div>
      <div>
        69
      </div>
      <div>
        70
      </div>
      <div>
        71
      </div>
      <div>
        72
      </div>
      <div>
        73
      </div>
      <div>
        74
      </div>
      <div>
        75
      </div>
      <div>
        76
      </div>
      <div>
        77
      </div>
      <div>
        78
      </div>
      <div>
        79
      </div>
      <div>
        80
      </div>
      <div>
        81
      </div>
      <div>
        82
      </div>
      <div>
        83
      </div>
      <div>
        84
      </div>
      <div>
        85
      </div>
      <div>
        86
      </div>
      <div>
        87
      </div>
      <div>
        88
      </div>
      <div>
        89
      </div>
      <div>
        90
      </div>
      <div>
        91
      </div>
      <div>
        92
      </div>
      <div>
        93
      </div>
      <div>
        94
      </div>
      <div>
        95
      </div>
      <div>
        96
      </div>
      <div>
        97
      </div>
      <div>
        98
      </div>
      <div>
        99
      </div>
      <div>
        100
      </div>
      <div>
        101
      </div>
      <div>
        102
      </div>
      <div>
        103
      </div>
      <div>
        104
      </div>
      <div>
        105
      </div>
      <div>
        106
      </div>
      <div>
        107
      </div>
      <div>
        108
      </div>
      <div>
        109
      </div>
      <div>
        110
      </div>
      <div>
        111
      </div>
      <div>
        112
      </div>
      <div>
        113
      </div>
      <div>
        114
      </div>
      <div>
        115
      </div>
      <div>
        116
      </div>
      <div>
        117
      </div>
      <div>
        118
      </div>
      <div>
        119
      </div>
      <div>
        120
      </div>
      <div>
        121
      </div>
      <div>
        122
      </div>
      <div>
        123
      </div>
      <div>
        124
      </div>
      <div>
        125
      </div>
      <div>
        126
      </div>
      <div>
        127
      </div>
      <div>
        128
      </div>
      <div>
        129
      </div>
      <div>
        130
      </div>
      <div>
        131
      </div>
      <div>
        132
      </div>
      <div>
        133
      </div>
      <div>
        134
      </div>
      <div>
        135
      </div>
      <div>
        136
      </div>
      <div>
        137
      </div>
      <div>
        138
      </div>
      <div>
        139
      </div>
      <div>
        140
      </div>
      <div>
        141
      </div>
      <div>
        142
      </div>
      <div>
        143
      </div>
      <div>
        144
      </div>
      <div>
        145
      </div>
      <div>
        146
      </div>
      <div>
        147
      </div>
      <div>
        148
      </div>
      <div>
        149
      </div>
      <div>
        150
      </div>
      <div>
        151
      </div>
      <div>
        152
      </div>
      <div>
        153
      </div>
      <div>
        154
      </div>
      <div>
        155
      </div>
      <div>
        156
      </div>
      <div>
        157
      </div>
      <div>
        158
      </div>
      <div>
        159
      </div>
      <div>
        160
      </div>
      <div>
        161
      </div>
      <div>
        162
      </div>
      <div>
        163
      </div>
      <div>
        164
      </div>
      <div>
        165
      </div>
      <div>
        166
      </div>
      <div>
        167
      </div>
      <div>
        168
      </div>
      <div>
        169
      </div>
      <div>
        170
      </div>
      <div>
        171
      </div>
      <div>
        172
      </div>
      <div>
        173
      </div>
      <div>
        174
      </div>
      <div>
        175
      </div>
      <div>
        176
      </div>
      <div>
        177
      </div>
      <div>
        178
      </div>
      <div>
        179
      </div>
      <div>
        180
      </div>
      <div>
        181
      </div>
      <div>
        182
      </div>
      <div>
        183
      </div>
      <div>
        184
      </div>
      <div>
        185
      </div>
      <div>
        186
      </div>
      <div>
        187
      </div>
      <div>
        188
      </div>
      <div>
        189
      </div>
      <div>
        190
      </div>
      <div>
        191
      </div>
      <div>
        192
      </div>
      <div>
        193
      </div>
      <div>
        194
      </div>
      <div>
        195
      </div>
      <div>
        196
      </div>
      <div>
        197
      </div>
      <div>
        198
      </div>
      <div>
        199
      </div>
    </div>
    <div
      class="mo-scrollBar__track mo-scrollBar__track--horizontal mo-scrollBar__track--hidden"
    >
      <div
        class="mo-scrollBar__thumb"
        style="width: calc(100% * 0.1);"
      />
    </div>
  </div>
</DocumentFragment>
`;

exports[`The ScrollBar Component Match Snapshot 1`] = `
<DocumentFragment>
  <div
    class="mo-scrollBar__wrapper"
  >
    <div
      class="mo-scrollBar__container mo-scrollBar__container--vertical"
    >
      <div>
        0
      </div>
      <div>
        1
      </div>
      <div>
        2
      </div>
      <div>
        3
      </div>
      <div>
        4
      </div>
      <div>
        5
      </div>
      <div>
        6
      </div>
      <div>
        7
      </div>
      <div>
        8
      </div>
      <div>
        9
      </div>
      <div>
        10
      </div>
      <div>
        11
      </div>
      <div>
        12
      </div>
      <div>
        13
      </div>
      <div>
        14
      </div>
      <div>
        15
      </div>
      <div>
        16
      </div>
      <div>
        17
      </div>
      <div>
        18
      </div>
      <div>
        19
      </div>
      <div>
        20
      </div>
      <div>
        21
      </div>
      <div>
        22
      </div>
      <div>
        23
      </div>
      <div>
        24
      </div>
      <div>
        25
      </div>
      <div>
        26
      </div>
      <div>
        27
      </div>
      <div>
        28
      </div>
      <div>
        29
      </div>
      <div>
        30
      </div>
      <div>
        31
      </div>
      <div>
        32
      </div>
      <div>
        33
      </div>
      <div>
        34
      </div>
      <div>
        35
      </div>
      <div>
        36
      </div>
      <div>
        37
      </div>
      <div>
        38
      </div>
      <div>
        39
      </div>
      <div>
        40
      </div>
      <div>
        41
      </div>
      <div>
        42
      </div>
      <div>
        43
      </div>
      <div>
        44
      </div>
      <div>
        45
      </div>
      <div>
        46
      </div>
      <div>
        47
      </div>
      <div>
        48
      </div>
      <div>
        49
      </div>
      <div>
        50
      </div>
      <div>
        51
      </div>
      <div>
        52
      </div>
      <div>
        53
      </div>
      <div>
        54
      </div>
      <div>
        55
      </div>
      <div>
        56
      </div>
      <div>
        57
      </div>
      <div>
        58
      </div>
      <div>
        59
      </div>
      <div>
        60
      </div>
      <div>
        61
      </div>
      <div>
        62
      </div>
      <div>
        63
      </div>
      <div>
        64
      </div>
      <div>
        65
      </div>
      <div>
        66
      </div>
      <div>
        67
      </div>
      <div>
        68
      </div>
      <div>
        69
      </div>
      <div>
        70
      </div>
      <div>
        71
      </div>
      <div>
        72
      </div>
      <div>
        73
      </div>
      <div>
        74
      </div>
      <div>
        75
      </div>
      <div>
        76
      </div>
      <div>
        77
      </div>
      <div>
        78
      </div>
      <div>
        79
      </div>
      <div>
        80
      </div>
      <div>
        81
      </div>
      <div>
        82
      </div>
      <div>
        83
      </div>
      <div>
        84
      </div>
      <div>
        85
      </div>
      <div>
        86
      </div>
      <div>
        87
      </div>
      <div>
        88
      </div>
      <div>
        89
      </div>
      <div>
        90
      </div>
      <div>
        91
      </div>
      <div>
        92
      </div>
      <div>
        93
      </div>
      <div>
        94
      </div>
      <div>
        95
      </div>
      <div>
        96
      </div>
      <div>
        97
      </div>
      <div>
        98
      </div>
      <div>
        99
      </div>
      <div>
        100
      </div>
      <div>
        101
      </div>
      <div>
        102
      </div>
      <div>
        103
      </div>
      <div>
        104
      </div>
      <div>
        105
      </div>
      <div>
        106
      </div>
      <div>
        107
      </div>
      <div>
        108
      </div>
      <div>
        109
      </div>
      <div>
        110
      </div>
      <div>
        111
      </div>
      <div>
        112
      </div>
      <div>
        113
      </div>
      <div>
        114
      </div>
      <div>
        115
      </div>
      <div>
        116
      </div>
      <div>
        117
      </div>
      <div>
        118
      </div>
      <div>
        119
      </div>
      <div>
        120
      </div>
      <div>
        121
      </div>
      <div>
        122
      </div>
      <div>
        123
      </div>
      <div>
        124
      </div>
      <div>
        125
      </div>
      <div>
        126
      </div>
      <div>
        127
      </div>
      <div>
        128
      </div>
      <div>
        129
      </div>
      <div>
        130
      </div>
      <div>
        131
      </div>
      <div>
        132
      </div>
      <div>
        133
      </div>
      <div>
        134
      </div>
      <div>
        135
      </div>
      <div>
        136
      </div>
      <div>
        137
      </div>
      <div>
        138
      </div>
      <div>
        139
      </div>
      <div>
        140
      </div>
      <div>
        141
      </div>
      <div>
        142
      </div>
      <div>
        143
      </div>
      <div>
        144
      </div>
      <div>
        145
      </div>
      <div>
        146
      </div>
      <div>
        147
      </div>
      <div>
        148
      </div>
      <div>
        149
      </div>
      <div>
        150
      </div>
      <div>
        151
      </div>
      <div>
        152
      </div>
      <div>
        153
      </div>
      <div>
        154
      </div>
      <div>
        155
      </div>
      <div>
        156
      </div>
      <div>
        157
      </div>
      <div>
        158
      </div>
      <div>
        159
      </div>
      <div>
        160
      </div>
      <div>
        161
      </div>
      <div>
        162
      </div>
      <div>
        163
      </div>
      <div>
        164
      </div>
      <div>
        165
      </div>
      <div>
        166
      </div>
      <div>
        167
      </div>
      <div>
        168
      </div>
      <div>
        169
      </div>
      <div>
        170
      </div>
      <div>
        171
      </div>
      <div>
        172
      </div>
      <div>
        173
      </div>
      <div>
        174
      </div>
      <div>
        175
      </div>
      <div>
        176
      </div>
      <div>
        177
      </div>
      <div>
        178
      </div>
      <div>
        179
      </div>
      <div>
        180
      </div>
      <div>
        181
      </div>
      <div>
        182
      </div>
      <div>
        183
      </div>
      <div>
        184
      </div>
      <div>
        185
      </div>
      <div>
        186
      </div>
      <div>
        187
      </div>
      <div>
        188
      </div>
      <div>
        189
      </div>
      <div>
        190
      </div>
      <div>
        191
      </div>
      <div>
        192
      </div>
      <div>
        193
      </div>
      <div>
        194
      </div>
      <div>
        195
      </div>
      <div>
        196
      </div>
      <div>
        197
      </div>
      <div>
        198
      </div>
      <div>
        199
      </div>
    </div>
    <div
      class="mo-scrollBar__track mo-scrollBar__track--vertical mo-scrollBar__track--hidden"
    >
      <div
        class="mo-scrollBar__thumb"
        style="height: calc(100% * 0.1);"
      />
    </div>
  </div>
</DocumentFragment>
`;
